File: uva10465/uva10465.py
from sys import stdin
import sys
sys.setrecursionlimit(10000000)
INF = float('inf')

def solve(t):
    global mem
    ans = None
    if t < 0:
        ans = -INF
    else:
        if mem[t] != -1:
            ans = mem[t]
        elif t == 0:
            ans = 0 
        else:
            ans = max(1+solve(t-m),1+solve(t-n))
        mem[t] = ans
    return ans
        
    



def main():
    global m,n,mem
    line = stdin.readline().strip()
    while len(line) and line != "" and line != "\n":
        m,n,t = map(int,line.split())
        mem,ok,i = [-1 for _ in range(t+1)],False,0
        ans = solve(t)
        if ans < 0:
            while i < t+1 and not ok:
                ok = mem[i] != -1
                i += 1
            print(solve(t-(i-1)),i-1)
        else:
            print(ans)
            
        #for i in range(t+1):
        #    if mem[i] != -1:
        #        print(i,end = " ")
        
        line = stdin.readline().strip()
        
main()      

File: uva10465/uva10465.cpp
#include <iostream>
#include <map>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <algorithm>
#define INF 99999999
#define ms(a,b) memset(a,b,sizeof(a))
using namespace std;

int n,m;
int mem[10001];

int solve(int t){
	int ans;
	if(t < 0)
		ans = -INF;
	else{
		if(mem[t] != -1)
			ans = mem[t];
		else if(t == 0)
			ans = 0;
		else
			ans = max(1+solve(t-n),1+solve(t-m));
		mem[t] = ans;

	}
	return ans;
}

int main(){

	int t,ans,i;
	bool ok;
	while(cin>>n>>m>>t){
		ok = false;
		ms(mem,-1);
		i = 0;
		ans = solve(t);
		if(ans < 0){
			while (i < t+1 && !ok){
				ok = mem[i] != -1;
				i += 1;
			}
			cout<<solve(t-(i-1))<<" "<<i-1<<endl;
		}
		else
			cout<<ans<<endl;

	}
	
	return 0;
}